src/serialize.hpp

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 Chia Network Inc
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//    http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SRC_SERIALIZE_HPP_
+#define SRC_SERIALIZE_HPP_
+
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+template<typename Type> class Serializable{
+public:
+    static void SerializeImpl(const Type& in, std::vector<uint8_t>& out)
+    {
+        size_t nSize = sizeof(in);
+        out.reserve(out.size() + nSize);
+        for (size_t i = 0; i < nSize; ++i) {
+            out.push_back(*((uint8_t*)&in + i));
+        }
+    }
+    static size_t DeserializeImpl(const std::vector<uint8_t>& in, Type& out, size_t position)
+    {
+        size_t size = sizeof(out);
+        if (position + size > in.size()) {
+            throw std::invalid_argument("DeserializeImpl: Trying to read out of bounds.");
+        }
+        for (size_t i = 0; i < size; ++i) {
+            *((uint8_t*)&out + i) = in[position + i];
+        }
+        return size;
+    }
+};
+
+template<typename TypeIn, typename TypeOut>
+void Serialize(const TypeIn& in, TypeOut& out)
+{
+    Serializable<TypeIn>::SerializeImpl(in, out);
+}
+
+template<typename TypeIn, typename TypeOut>
+size_t Deserialize(const TypeIn& in, TypeOut& out, const size_t position)
```

Comment:
similarly, `TypeIn` must be `std::vector<uint8_t> const&` afaict. It should be spelled out here for simplicity